### ISSUE TITLE
Ordering the dataset returned by calculate/develop

### ISSUE THREAD

**Issue by Mylleranton** [dbt-labs/dbt_metrics#130]
### Is this a new bug in dbt_metrics?

- [X] I believe this is a new bug in dbt_metrics
- [X] I have searched the existing issues, and I could not find an existing issue for this bug

### Current Behavior

Amazon Redshift returns the following error: 

`ERROR: expression_tree_walker unrecognized node type: 407`

when previewing (or running) model consisting solely of:

```sql
select * 
from {{ metrics.calculate(
    metric('new_user_registrations'),
    grain='month',
    dimensions=['subregion', 'country']) }}
```

### Expected Behavior

I expect both running and previewing a metric to work from within a model selecting from it.

### Steps To Reproduce

0. Set up a dbt Environment with an Amazon Redshift target
1. Set up a dummy seed table `dim_users.csv` to have some data to work with, containing
```csv 
user_id,country,subregion,user_registered_at
1,Norway,Nordics,2022-01-01
2,Norway,Nordics,2022-01-01
3,Norway,Nordics,2022-06-01
4,Norway,Nordics,2022-06-01
5,Norway,Nordics,2022-07-01
6,Sweden,Nordics,2022-01-01
7,Sweden,Nordics,2022-01-01
8,Sweden,Nordics,2022-06-01
9,Sweden,Nordics,2022-07-01
10,Sweden,Nordics,2022-07-01
```
2. Define the metric as follows:
```yaml
- name: new_user_registrations
    label: New User Registrations
    description: "The number of users that registered in the given time period."
    model: ref('dim_users')

    type: count
    sql: user_id

    timestamp: user_registered_at
    time_grains: [day, week, month, year]
    dimensions:
      - country
      - subregion
```

3. Create an example model and populate it with the malfunctioning sql:
```sql 
select * 
from {{ metrics.calculate(
    metric('new_user_registrations'),
    grain='month',
    dimensions=['subregion', 'country']) }}
```

4. Initialise all seeds and run dbt to get the error:
```
dbt seed --full-refresh
dbt run
```

### Relevant log output

```shell
Opening a new connection, currently in state init
 Redshift adapter: Connecting to Redshift using 'database' credentials
 Postgres adapter: Postgres error: expression_tree_walker unrecognized node type: 407
 finished collecting timing info
 On rpc.***.request: Close
Got an exception: Database Error
  expression_tree_walker unrecognized node type: 407
Got exception RPCException(10003, Database Error, {'type': 'DatabaseException', 'message': 'Database Error in rpc request (from remote system)\n  expression_tree_walker unrecognized node type: 407', 'raw_sql':...
```

### Environment

```markdown
- dbt-adapter & version: dbt-redshift v1.2.0
- dbt_metrics version: v0.3.6
```

### Which database adapter are you using with dbt?

redshift

### Additional Context

I examined some Google results regarding the specific Postgres-error 407 that was thrown, and stumbled on some information about ordering causing it in [some cases](https://stackoverflow.com/questions/57856121/redshift-unrecognized-node-type-407), so I examined the dbt macro code for `calculate`, which generates this sequence at the end (of the specific example) of the `__final` CTE:
```sql 
order by 1 desc, 2 desc, 3 desc
```

I tried manually executing the complied SQL outside of the dbt environment and directly in the AWS Redshift workbench and got the same error - **but when I removed the `order by 1 desc, 2 desc, 3 desc`, the query worked 100% as expected and returned the correct results.**

**Comment by Mylleranton** [comment]
After trying the steps to reproduce locally, I did not encounter the issue - but there seem to be some general issues with doing ordering from within CTEs, and it is a generally accepted anti-pattern for SQL:
**Don't:**
```sql 
select * from (select * from exampletable order by 1) sub
```
**Do:** 
If ordering is needed (which, in the case of dbt-metrics case seems to be more of a convenience than a requirement), perform ordering in the outermost/final SQL statement.

Therefore, I would suggest discussing whether the ordering should be kept or not (i.e. if it's there for some functional requirement or not), and it it should be kept - whether it can be moved to the outermost sql statement to avoid intermittent issues like these.

**Comment by callum-mcdata** [comment]
@Mylleranton Interesting! Looks like the fix here is moving the order by into the final select statement.

We recently re-added the `order by` logic in #119 . This was based on feedback from beta users of the dbt Semantic Layer that expressed mild frustration with unordered logic being returned. We decided to add the ordering logic (which actually brought it back because it had previously been in v0.1.0).

I've created https://github.com/dbt-labs/dbt_metrics/pull/131 to fix this in the 0.3.x series and will update the pre-release. 

The weird part is that we have testing for redshift and none of that triggered errors when we added the `order_by` logic. I wonder what triggered it in these circumstances that didn't get triggered in our testing.

**Comment by callum-mcdata** [comment]
@Mylleranton do you mind testing this out on `0.3.latest` by installing that branch with the following code:

```
packages:
  - git: "https://github.com/dbt-labs/dbt_metrics.git"
    revision: 0.3.latest
```

If that works we can push it to the 0.3.x version and then follow it up with an update to the 1.3.0 prerelease.

**Comment by Mylleranton** [comment]
@callum-mcdata Thanks for circling back so soon & explaining the reasoning behind the current implementation. 

After testing the `0.3.6` and `0.3.latest` revisions, I can confirm that I still have the issue with `0.3.6` (as expected) and that **the issue is gone in the `0.3.latest` revision (and the result set is still ordered as expected)!**

Seeing as I was unable to reproduce the error locally with my own steps to reproduce, and having the error only occur in our production server with a bigger table (1M+ rows with 30+ columns), my initial thought was that it might have to do with leader/slave node processing logic for Redshift, but that feels like a long shot... Might be worth doing some more digging on the root cause of the error message - but then again - the error likely was not widespread since I was unable to reproduce it myself.

Thanks for the rapid resolution!

**Comment by callum-mcdata** [comment]
@Mylleranton glad to know that this worked! Unfortunately I think I'm going to need to revert that change. 

We had a discussion internally and realized that this introduces more issues then it solves by adding the `order by` outside of the generated sql. The sql:

```
select *
from {{metrics....}}
```

implies that the generated statement is a subquery that can have additional manipulations applied to it like `where` or `group`, etc. Unfortunately that is impossible if we move the `order by` to the section after the generated sql. So in reality its closer to:

```
select *
from {{metrics...}}
order by 1,2,3....
```

Our proposed fix is:
1.  removing ordering from redshift for the time being
2. Try wrapping the entire statement in () and having the order be after another nested select. Would you be willing to test this out?

**Comment by Mylleranton** [comment]
@callum-mcdata 
Hm, that reasoning is solid, and I completely overlooked the fact that the `order by`-clause in the outermost select completely restricts what you are able to do with the `select * from {{metrics...}}`. So for that alone, I'm on board with reverting the temporary fix.

To be honest, I have done a fair bit of research but have not been able to learn that much more about the core issue that triggers this error message. **At this point in time, I am also unable to replicate the issue - even on the same cluster I had it on originally**, which again points to it being very intermittent and seemingly random. Today, I get the same results using compiled code from `0.3.6` and `0.3.latest`.

Moving forward, I guess that there are a few options on how to address this. 
**0. Do nothing, as this seems to be very intermittent and random:** A valid option at this point in time as the number of affected users likely is very small but runs the risk of having to revisit this at a later point in time.
**1. Removing the ordering from select adapters:** This seems like an easy fix to alleviate this particular issue alone but does not address the larger issue of sql best practices/any (potential) future issues. Rather than making the choice on the behalf of the users, could the result set ordering not be configurable by the user instead? A global config flag for the package should suffice.
**2. Wrapping the select in () and shuffle the order by there:** Also a valid point, but not sure if that would help. Since I'm unable to reproduce the error now, I'm afraid I cannot test it to validate it resolves the issue...

Personally (without knowing the discussions you've had internally & here), I'd probably opt for option 1, i.e. making the ordering configurable because of a few reasons:
1. It will allow those that need the ordering for manual inspection/convenience to have it there, and it will allow those that do not need it to not have it (and maybe earn marginally better performance).
2. It will not paint the package itself into a corner with some functionality available on certain adapters only with the added maintenance & complexity it results in
3. It will leave the choice up to the user to configure, which is generally considered good practice for optional/convenience features and it will likely reduce the risk/amount of bug-fixing/maintenance issues related to ordering.

**Comment by callum-mcdata** [comment]
@Mylleranton unfortunately we won't have time to release a config flag before Coalesce when the next version of dbt-core is going to go out. I definitely agree that the longer term best plan is for there to be some method of configuration with a reasonable default.

In the meantime I've gone forward with option 1 via PR #136 but I will keep this issue open until a more holistic solution is implemented.

**Comment by Mylleranton** [comment]
@callum-mcdata Understandable, thanks for that good discussion & resolution for now!

**On the topic of a more holistic solution:**
I guess one could leverage the config framework that was shipped in the `1.3.0-rc2`, but given the (subtle) differences between `per-metric` configs and `per-project` configs, it might not be that easy.

Given that ordering takes place in `gen_final_cte` which acts on the entire list of metrics (contrary to `gen_secondary_calculation` (#96) which implemented configs in the first place), it could contain multiple conflicting config values if we allow a `per-metric` configuration. At the same time, there might be use cases for when one would want some metrics ordered and some not - but that would then require some form or precedence (for example *"if one metric is configured to be ordered, then any expression where it's part will be ordered"*) in the config should the metric list contain many metrics. 

As a first step, a `per-project` configuration might be a good choice to keep things simple. And if things are made configurable - I guess *what* should be configurable needs to be discussed. Is it a simple boolean `[True, False]` or do we allow the ordering to be specified: `['desc', 'asc', 'None']`?

Happy to discuss further if needed when/if this gets picked up.

**Comment by callum-mcdata** [comment]
Let's take a step back even further and tackle it from first principles: **What is the value provided by ordering the datasets returned to end users?**

I would agree that ordering by the `date` column has value because it sorts the metrics into a linear range. Any ordering after that though ... I'm not convinced 🤔 . What do end-users **gain** by sorting their dimension values?

**Important Distinction**: In any metric query that has a time grain not equal to `all_time`, we would not want to support ordering on the metric value itself. If customers want to perform ordering at that level, I'd argue that it should be pushed down to the end query and not a function of the metric calculation.
